controllers/nnf_port_manager_controller.go

```diff
@@ -0,0 +1,296 @@
+			log.Info("Allocation unused", "requester", status.Requester, "status", status.Status)
+			if status.Status == nnfv1alpha1.NnfPortManagerAllocationStatusInUse {
+				status.Requester = nil
+				status.Status = nnfv1alpha1.NnfPortManagerAllocationStatusFree
+			} else if status.Status != nnfv1alpha1.NnfPortManagerAllocationStatusFree {
+				failedIndices = append(failedIndices, idx)
+			}
+		}
+	}
+
+	for idx := range failedIndices {
+		failedIdx := failedIndices[len(failedIndices)-1-idx] // remove in reverse order
+		mgr.Status.Allocations = append(mgr.Status.Allocations[:failedIdx], mgr.Status.Allocations[failedIdx+1:]...)
+	}
+}
+
+func (r *NnfPortManagerReconciler) findAllocationStatus(mgr *nnfv1alpha1.NnfPortManager, spec AllocationSpec) *AllocationStatus {
+	for idx := range mgr.Status.Allocations {
+		status := &mgr.Status.Allocations[idx]
+		if status.Requester == nil {
+			continue
+		}
+
+		if *status.Requester == spec.Requester {
+			return status
+		}
+	}
+
+	return nil
+}
+
+// isAllocated returns true if the provided specification is in the Port Manager's allocation
+// status', and false otherwise.
+func (r *NnfPortManagerReconciler) isAllocated(mgr *nnfv1alpha1.NnfPortManager, spec AllocationSpec) bool {
+	return r.findAllocationStatus(mgr, spec) != nil
+}
+
+// Find free ports to satisfy the provided specification.
+func (r *NnfPortManagerReconciler) findFreePorts(log logr.Logger, mgr *nnfv1alpha1.NnfPortManager, config *dwsv1alpha1.SystemConfiguration, spec AllocationSpec) ([]uint16, nnfv1alpha1.NnfPortManagerAllocationStatusStatus) {
+
+	portsInUse := make([]uint16, 0)
+	for _, status := range mgr.Status.Allocations {
+		if status.Status == nnfv1alpha1.NnfPortManagerAllocationStatusInUse {
+			portsInUse = append(portsInUse, status.Ports...)
+		}
+	}
+
+	isPortInUse := func(port uint16) bool {
+		for _, p := range portsInUse {
+			if p == port {
+				return true
+			}
+		}
+
+		return false
+	}
+
+	count := spec.Count
+	ports := make([]uint16, 0)
+
+	// We first turn to the system configuration and try to allocate free ports from there.
+	itr := portsutil.NewPortIterator(config.Spec.Ports)
+	for port := itr.Next(); port != portsutil.InvalidPort; port = itr.Next() {
+
+		if isPortInUse(port) {
+			continue
+		}
+
+		ports = append(ports, port)
+
+		if len(ports) >= count {
+			break
+		}
+	}
+
+	if len(ports) >= count {
+		log.Info("Ports claimed from system configuration", "ports", ports)
+		return ports[:count], nnfv1alpha1.NnfPortManagerAllocationStatusInUse
+	}
+
+	// If we still haven't found a sufficient number of free ports, free up unused allocations
+	// in the status' until we have the desired number of ports.
+
+	const (
+		exhausted = iota
+		more
+	)
+
+	// Search for a free allocation and claim ports from that allocation. Returns 'more' if there
+	// are potentially more allocations available with free ports, and 'exhausted' otherwise.
+	claimPortsFromFreeAllocation := func() int {
```

Comment:
Couldn't you just return a bool here?

Reply:
Then you get into the argument of what your return value represents. 

Does `claimPortsFromFreeAllocation() == true` mean I've claimed the ports I need? Does it mean I've exhausted the ports and failed?

With an enum like this, I've effectively implemented named bools.